tesseract/tesseract_planning/tesseract_motion_planners/include/tesseract_motion_planners/descartes/impl/descartes_positioner_sampler.hpp

```diff
@@ -0,0 +1,206 @@
+  , tool_pose_sampler_(tool_pose_sampler)
+  , positioner_kinematics_(std::move(positioner_kinematics))
+  , robot_kinematics_(std::move(robot_kinematics))
+  , collision_(std::move(collision))
+  , world_to_positioner_base_(current_state->transforms.at(positioner_kinematics_->getBaseLinkName()))
+  , world_to_robot_base_(current_state->transforms.at(robot_kinematics_->getBaseLinkName()))
+  , positioner_limits_(positioner_kinematics_->getLimits())
+  , positioner_sample_resolution_(std::move(positioner_sample_resolution))
+  , robot_tcp_(robot_tcp)
+  , robot_reach_(robot_reach)
+  , allow_collision_(allow_collision)
+  , dof_(positioner_kinematics_->numJoints() + robot_kinematics_->numJoints())
+  , ik_seed_(Eigen::VectorXd::Zero(dof_))
+  , is_valid_(is_valid)
+{
+}
+
+template <typename FloatType>
+bool DescartesPositionerSampler<FloatType>::sample(std::vector<FloatType>& solution_set)
+{
+  double distance = std::numeric_limits<double>::min();
+  int num_joints = static_cast<int>(positioner_kinematics_->numJoints());
+  std::vector<Eigen::VectorXd> dof_range;
+  dof_range.reserve(static_cast<std::size_t>(num_joints));
+  for (int dof = 0; dof < num_joints; ++dof)
+  {
+    int cnt = std::ceil(std::abs(positioner_limits_(dof, 1) - positioner_limits_(dof, 0)) /
+                        positioner_sample_resolution_(dof));
+    dof_range.push_back(Eigen::VectorXd::LinSpaced(cnt, positioner_limits_(dof, 0), positioner_limits_(dof, 1)));
+  }
+
+  tesseract_common::VectorIsometry3d tool_poses = tool_pose_sampler_(tool_pose_);
+  for (std::size_t i = 0; i < tool_poses.size(); ++i)
+  {
+    // Tool pose subtract robot tcp
+    Eigen::Isometry3d tool_pose = tool_poses[i] * robot_tcp_.inverse();
+    Eigen::Matrix<FloatType, Eigen::Dynamic, 1> positioner_pose(num_joints);
+    nested_ik(0, dof_range, tool_pose, positioner_pose, solution_set, false, distance);
+  }
+
+  if (solution_set.empty() && allow_collision_)
+    getBestSolution(solution_set, tool_poses, dof_range);
+
+  return !solution_set.empty();
+}
+
+template <typename FloatType>
+bool DescartesPositionerSampler<FloatType>::isCollisionFree(const FloatType* vertex)
+{
+  if (collision_ == nullptr)
+    return true;
+  else
+    return collision_->validate(vertex, dof_);
+}
+
+template <typename FloatType>
+void DescartesPositionerSampler<FloatType>::nested_ik(
+    const int loop_level,
+    const std::vector<Eigen::VectorXd>& dof_range,
+    const Eigen::Isometry3d& p,
+    Eigen::Ref<Eigen::Matrix<FloatType, Eigen::Dynamic, 1>> sample_pose,
+    std::vector<FloatType>& solution_set,
+    const bool get_best_solution,
+    double& distance)
+{
+  if (loop_level >= positioner_kinematics_->numJoints())
+  {
+    ikAt(p, sample_pose, solution_set, get_best_solution, distance);
+    return;
+  }
+
+  for (long i = 0; i < static_cast<long>(dof_range[static_cast<std::size_t>(loop_level)].size()); ++i)
+  {
+    sample_pose(loop_level) = dof_range[static_cast<std::size_t>(loop_level)][i];
```

Comment:
Is this doing what you intended? Could we get some comments in some of the more complicated spots explaining what exactly is going on?

Reply:
I added documentation and it should also be adding +1 to the count to get the correct results from LinSpaced. It is fixed.